synapse/storage/persist_events.py

```diff
@@ -583,7 +583,8 @@ async def _persist_event_batch(
                 current_state_for_room=current_state_for_room,
                 state_delta_for_room=state_delta_for_room,
                 new_forward_extremeties=new_forward_extremeties,
-                backfilled=backfilled,
+                use_negative_stream_ordering=backfilled,
+                inhibit_local_membership_updates=backfilled,
```

Comment:
> If there are ways of grouping some of these flags together into coherent bits of functionality, I think that would be better.
>
> https://github.com/matrix-org/synapse/pull/11396#pullrequestreview-811548838

Around that idea, there is a nicety around grouping up the behavior under the `backfilled` label.

Somewhere at the multiple top-levels when we refactor this further, we're going to have to make sure the `backfilled` logic is consistent.


Wdyt about having a static lookup map:

```py

class BackfilledAttributes:
    use_negative_stream_ordering = False
    inhibit_local_membership_updates = False
    other_more_condition = True
    extraneous_more_condition = False
    different_more_condition = True
```

And it can be used like this. I don't know. I'm not convinced of this usage and haven't figured a good clean way for the `else ...` spot (whether normal events have their own` NormalAttributes` or just be to negate the backfill one) but it seems close. Something in this vein.

```py
await self.persist_events_store._persist_events_and_state_updates(
    chunk,
    current_state_for_room=current_state_for_room,
    state_delta_for_room=state_delta_for_room,
    new_forward_extremeties=new_forward_extremeties,
    use_negative_stream_ordering=BackfilledAttributes.use_negative_stream_ordering if backfilled else ...,
    inhibit_local_membership_updates=BackfilledAttributes.inhibit_local_membership_updates if backfilled else ...,
    other_more_condition=BackfilledAttributes.other_more_condition if backfilled else ...,
    extraneous_more_condition=BackfilledAttributes.extraneous_more_condition if backfilled else ...,
    different_more_condition=BackfilledAttributes.different_more_condition if backfilled else ...,
)
```

Reply:
There's something to be said for identifying different "classes" of event (in this case, 'backfilled' and, uh, 'normal') with a list of exactly how those classes differ in behaviour.  In that case you can just pass the 'class' identifier around rather than a sea of booleans.

We do something pretty similar with room versions (see https://github.com/matrix-org/synapse/blob/v1.47.0/synapse/api/room_versions.py#L50 etc) and I don't hate it.

That said, I'm yet to be convinced that such a solution isn't over-engineered in this case - if we've only really got a few degrees of freedom here, then booleans for them might be easier. Can I reserve judgement on this?

Reply:
Great comparison to `RoomVersion` 👌

> Can I reserve judgement on this?

Totally fine to push this to when we have more a need. Just wanted to throw out the inkling of an idea that the original `backfilled` had some benefit and how we could retain some of that in this new refactor ⏩